### DISCUSSION TITLE
AppwriteException: Missing required parameter: "data"

### DISCUSSION THREAD

**Discussion by kirtanchandak** [appwrite/appwrite#5027] [Q&A] [answered]
### AppwriteException: Missing required parameter: "data"
```javascript
const handleSumbit = async (e) => {
    e.preventDefault();
    const promise = databases.createDocument("ID", uuidv4(), {
      todo,
    });

    promise.then(
      function (response) {
        console.log(response);
      },
      function (error) {
        console.log(error);
      }
    );
    e.target.reset();
```

**Comment by stnguyen90** [comment, marked as answer]
@kirtanchandak, assuming you're using version 10.2.0 of the [appwrite](https://www.npmjs.com/package/appwrite) package, there are 4 required parameters, but you're only passing 3.

```javascript
async createDocument<Document extends Models.Document>(databaseId: string, collectionId: string, documentId: string, data: Omit<Document, keyof Models.Document>, permissions?: string[]): Promise<Document>
```

Source: [sdk-for-web](https://github.com/appwrite/sdk-for-web/blob/1dde5c7bc91b18bfc1f4e2a80b0ba77017db66fe/src/services/databases.ts#L63)

FYI, you can use [`ID.unique()`](https://github.com/appwrite/sdk-for-web/blob/1dde5c7bc91b18bfc1f4e2a80b0ba77017db66fe/src/id.ts#L6) for a unique ID.

**Reply by sandesh682** [reply]
> ```js
> data: Omit<Document, keyof Models.Document>
> ```

You can use below code:

    const promise = databases.createDocument(
      "dbId",
      "collectionId",
      uuidv4(),
      {
        todo,
      },
      undefined
    );

**Reply by himankshu1** [reply]
I am getting the same error for userId. However, the records are getting added to the collection but still I am getting this error

import { client, databases, storage, appwriteConfig } from "./config";
import { ID } from "appwrite";

export const saveUserToDB = async ({ name, email, username, userId, imageUrl }) => {
    try {
        console.log(name, email, username, userId);
        return await databases.createDocument(
            appwriteConfig.databaseId,
            appwriteConfig.userCollectionId,
            ID.unique(),
            { name, email, username, userId, imageUrl }
        );
    } catch (error) {
        console.log("Error while creating a user in 'user' database, saveUserToDB: ", error);
    }
}

**Comment by kirtanchandak** [comment]
### Actually it takes 5 parameters 
### But not I am getting an new error - AppwriteException: Invalid document structure: Unknown attribute: "0"
```
const handleSumbit = async (e) => {
    e.preventDefault();
    const promise = databases.createDocument(
      "ID",
      "ID",
      uuidv4(),
      [todo],
      [
        Permission.read(Role.any()),
        Permission.update(Role.any()),
        Permission.update(Role.any()),
      ]
    );

    promise.then(
      function (response) {
        console.log(response);
      },
      function (error) {
        console.log(error);
      }
    );
    e.target.reset();
  };
```

**Reply by stnguyen90** [reply]
Answered in the related discussion: https://github.com/appwrite/appwrite/discussions/5029#discussioncomment-4783306
